server/tests/src/main/scala/sttp/tapir/server/tests/ServerWebSocketTests.scala

```diff
@@ -141,7 +204,116 @@ abstract class ServerWebSocketTests[F[_], S <: Streams[S], OPTIONS, ROUTE](
         .send(backend)
         .map(_.body shouldBe Left("Not a WS!"))
     }
-  )
+  ) ++ autoPingTests ++ failingPipeTests ++ handlePongTests
```

Comment:
in fact we have a test for non-ws request to a ws endpoint above, so maybe this works, just would need extending with a status code check

still a test for a ws request to a non-ws endpoint could be useful

Reply:
We have a test:
```scala
  testServer(
      endpoint
        .in(isWebSocket)
        .errorOut(stringBody)
        .out(stringWs),
      "non web-socket request"
    )(isWS => if (isWS) pureResult(stringEcho.asRight) else pureResult("Not a WS!".asLeft)) { (backend, baseUri) =>
```

does this mean that the user has to explicitly handle this?

Reply:
Yes and no ;) This test probably needs improving / commenting: I think it might test for the case of a "dynamic" web socket output - that is, we can decide dynamically if we return a web socket, or not. 

The endpoint above works for both cases, deciding in the server logic what type of output to produce. Using `isWebSocket` is probably just me being lazy, and the test should use a query parameter or sth like this, and be split into four: with/without WS upgrade headers, with/without the "correct" query parameter value.

This would also test for the case when we want to do some validation of the request, before returning a web socket - e.g, testing security - which case we want to return a nice error to the user